core/state_processor.go

```diff
@@ -205,17 +206,41 @@ func (p *StateProcessor) Process(block *types.Block) (types.Receipts, []*types.L
 			return nil, nil, nil, 0, fmt.Errorf("could not apply tx %d [%v]: %w", i, tx.Hash().Hex(), err)
 		}
 		statedb.Prepare(tx.Hash(), i)
-		receipt, err := applyTransaction(msg, p.config, p.hc, nil, gp, statedb, blockNumber, blockHash, tx, usedGas, vmenv)
-		if err != nil {
-			return nil, nil, nil, 0, fmt.Errorf("could not apply tx %d [%v]: %w", i, tx.Hash().Hex(), err)
+		var receipt *types.Receipt
+		if tx.Type() == types.ExternalTxType {
+			if _, exists := etxSet[tx.Hash()]; !exists { // Verify that the ETX exists in the set
+				return nil, nil, nil, 0, fmt.Errorf("invalid external transaction: etx %x not found in unspent etx set", tx.Hash())
+			}
+			prevZeroBal := prepareApplyETX(statedb, tx)
+			receipt, err = applyTransaction(msg, p.config, p.hc, nil, gp, statedb, blockNumber, blockHash, tx, usedGas, vmenv)
+			if err == nil {
+				statedb.AddBalance(common.ZeroAddr, prevZeroBal) // Add previous zero address balance to residual zero address balance, even if the transaction was unsuccessful (e.g. failed)
+			}
+
+			if err != nil {
+				statedb.SetBalance(common.ZeroAddr, prevZeroBal) // In the case of an error, reset the balance to what it previously was
+				return nil, nil, nil, 0, fmt.Errorf("could not apply tx %d [%v]: %w", i, tx.Hash().Hex(), err)
+			}
+
+			delete(etxSet, tx.Hash()) // This ETX has been spent so remove it from the unspent set
```

Comment:
Unless I am missing something, a failure at that `p.engine.Finalize()` step would result in the entire block getting rejected, rather than the block being accepted with a failed TX.

If the block is rejected, then no change will happen to the ETX set, because the ETX set is always re-loaded by parent hash, before processing a new block.